packages/react-reconciler/src/ReactMemoCache.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {MemoCache} from 'shared/ReactTypes';
+import type {Dispatcher} from './ReactInternalTypes';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+const {ReactCurrentDispatcher} = ReactSharedInternals;
+
+export const MEMO_CACHE_UNSET_SENTINEL: symbol = Symbol.for(
+  'react.usememocache_sentinel',
+);
+
+/**
+ * Implementation of `useMemoCache()` return type
+ */
+export default class MemoCacheImpl implements MemoCache {
+  data: Array<any>;
+
+  constructor(data: Array<any>) {
+    this.data = data;
+  }
+
+  get size(): number {
+    return this.data.length;
+  }
+
+  has(index: number): boolean {
+    if (index >= this.data.length) {
+      throw new Error('Memo cache index of out bounds');
+    }
+    return this.data[index] !== MEMO_CACHE_UNSET_SENTINEL;
+  }
+
+  get(index: number): any {
+    if (index >= this.data.length) {
+      throw new Error('Memo cache index of out bounds');
+    }
+    const value = this.data[index];
+    if (value === MEMO_CACHE_UNSET_SENTINEL) {
+      throw new Error('Memo cache index accessed before being initialized');
+    }
+    return value;
+  }
+
+  set(index: number, value: any) {
+    if (index >= this.data.length) {
+      throw new Error('Memo cache index of out bounds');
+    }
+    this.data[index] = value;
+  }
+
+  reset(): void {
+    this.data.fill(MEMO_CACHE_UNSET_SENTINEL);
+  }
+
+  withDispatcher_TEMPORARY_DEBUG_API_DO_NOT_USE<T>(
+    dispatcher: Dispatcher,
+    fn: () => T,
+  ): T | void {
+    const previousDispatcher = ReactCurrentDispatcher.current;
+    ReactCurrentDispatcher.current = dispatcher;
+    let value;
+    try {
+      value = fn();
+    } finally {
+      ReactCurrentDispatcher.current = previousDispatcher;
+    }
+    return value;
```

Comment:
You could just directly return, unless Flow doesn't like that?
```suggestion
    try {
      return fn();
    } finally {
      ReactCurrentDispatcher.current = previousDispatcher;
    }
```